### ISSUE TITLE
Flower keeps busy IMHO unnecessarily in CPU and IO: needs further investigation

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad-registry#301]
I think this needs further investigation since it remains the busiest process whenever there seems no activity otherwise.

CPU -- spikes into 100% averaging to ~60%

```shell
root@typhon:/home/yoh/proj/dandi/zarr-benchmarking/tools# pidstat -p 64122 5
Linux 6.1.0-17-amd64 (typhon)   02/13/2024      _x86_64_        (32 CPU)

09:22:33 AM   UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
09:22:38 AM  1036     64122   52.60    3.80    0.00    0.00   56.40     7  celery
09:22:43 AM  1036     64122   71.60    4.00    0.00    0.00   75.60     4  celery
09:22:48 AM  1036     64122   52.60    3.40    0.00    0.20   56.00    18  celery
09:22:53 AM  1036     64122   52.20    3.20    0.00    0.00   55.40     3  celery

root@typhon:/home/yoh/proj/dandi/zarr-benchmarking/tools# pidstat -p 64122 1
Linux 6.1.0-17-amd64 (typhon)   02/13/2024      _x86_64_        (32 CPU)

09:22:55 AM   UID       PID    %usr %system  %guest   %wait    %CPU   CPU  Command
09:22:56 AM  1036     64122   26.00    2.00    0.00    0.00   28.00     2  celery
09:22:57 AM  1036     64122   81.00    5.00    0.00    0.00   86.00    18  celery
09:22:58 AM  1036     64122   27.00    2.00    0.00    0.00   29.00     3  celery
09:22:59 AM  1036     64122  100.00    0.00    0.00    0.00  100.00     3  celery
09:23:00 AM  1036     64122   57.00    8.00    0.00    0.00   65.00     3  celery
09:23:01 AM  1036     64122   73.00    7.00    0.00    0.00   80.00     3  celery
09:23:02 AM  1036     64122   25.00    2.00    0.00    0.00   27.00    18  celery
```

and according to `iotop` it writes at about `15MBps` on average:

```
Total DISK READ:         0.00 B/s | Total DISK WRITE:        15.91 M/s
Current DISK READ:       0.00 B/s | Current DISK WRITE:      17.11 M/s
    TID  PRIO  USER     DISK READ DISK WRITE>    COMMAND                                                                                                                                                                                                           
  64122 be/4 ito         0.00 B/s   15.91 M/s python3 /venv/bin/celery -A datalad_registry.make_celery:celery_app flower
```

what is odd is that process tries to access non-existing `/data/flower` once in awhile:

```
root@typhon:/home/yoh/proj/dandi/zarr-benchmarking/tools# ls -l /proc/64122/fd/20
ls: cannot access '/proc/64122/fd/20': No such file or directory
root@typhon:/home/yoh/proj/dandi/zarr-benchmarking/tools# ls -l /proc/64122/fd/20
lrwx------ 1 ito ito 64 Feb 13 09:25 /proc/64122/fd/20 -> /data/flower
root@typhon:/home/yoh/proj/dandi/zarr-benchmarking/tools# ls -l /data/flower
ls: cannot access '/data/flower': No such file or directory
```
Have a clue what is that about @candleindark ?

**Comment by candleindark** [comment]
![Screenshot 2024-02-13 at 1 33 42 PM](https://github.com/datalad/datalad-registry/assets/12135617/6fb5c48b-be76-4e67-b458-080fe6ec789f)

As we discussed in our weekly meeting, we come to the conclusion that the `flower` database file must be repeatedly rewritten in a very short period time. As illustrated in the above screenshot, we can see that the size of the `flower` file oscillates between 29M to 16K in a very short period of time.

This issue is basically the continuation of #258, and since our requests to address this issue on https://github.com/mher/flower/issues/771 are not yet answered by the Flower team for two months. We need to provide a solution to address this problem ourselves without further waiting.

**Comment by yarikoptic** [comment]
it is odd indeed... let's at least make it in memory FS. I am ready - there is `/data/srv/registry/services/monitor/data2` to try on. Let's bring down that registry tomorrow, I will change it to become `/data` instead of `/data2` and we will restart the service.

**Comment by yarikoptic** [comment]
what is that `data/` anyways ? looking at https://github.com/mher/flower/pull/1111/files#diff-aef274fd2ef81ae4ec38d9c95b94da974653a6b76a5367eee2b180894de1a423L95, it seems that flower by default `persistent=False` and ` state_save_interval=0,` if that is about that DB, so should have not saved it at all ?

**Comment by candleindark** [comment]
> it is odd indeed... let's at least make it in memory FS. I am ready - there is `/data/srv/registry/services/monitor/data2` to try on. Let's bring down that registry tomorrow, I will change it to become `/data` instead of `/data2` and we will restart the service.

The only thing in the `data` directory is the `flower` file. The `flower` file is the one that stores all the records of the tasks. The records that we see in the monitor panel as depicted below.

![Screenshot 2024-02-13 at 8 14 55 PM](https://github.com/datalad/datalad-registry/assets/12135617/958930b3-959c-4ece-8845-cd70cdf2b9e9)

This file is set by the `db` setting of flower, https://flower.readthedocs.io/en/latest/config.html#db.

We set this file to be `/data/flower` inside the container, https://github.com/datalad/datalad-registry/blob/1a94c920c2a4047a8b9bf4ceddaaeabb91806090/docker-compose.dev.yml#L87.

If the memory FS solution works, it will be better than turning the [persistent mode](https://flower.readthedocs.io/en/latest/config.html#persistent) in Flower off.

**Comment by yarikoptic** [comment]
but my question is either that "persistent mode" is on and then how it is on if by default it is off?

**Comment by candleindark** [comment]
I set the persistent mode on because I want the record of tasks even after Flower restart, https://github.com/datalad/datalad-registry/blob/1a94c920c2a4047a8b9bf4ceddaaeabb91806090/docker-compose.dev.yml#L86. It has been always on since I have Flower set up. We can try having the persistent mode turned off.

**Comment by candleindark** [comment]
With the memory FS setup, the CPU usage situation is even worse. I have set the persistent mode to off, the default setting. Let's see what the CPU usage with that setting.

**Comment by candleindark** [comment]
After setting the persistent mode to off, the default. We no longer have the high CPU usage problem.

![Screenshot 2024-02-19 at 10 02 09 AM](https://github.com/datalad/datalad-registry/assets/12135617/d2b0132a-c0dd-425c-adbf-ca76ba88f17b)

Above is a snapshot after 3 days since the last start of Datalad-Registry.

@yarikoptic Let me know if you want to keep the persistent mode on. That will require modifying Flower's code, and we can attempt that ourselves.

I have already reported our new finding about this problem to the Flower team at https://github.com/mher/flower/issues/771#issuecomment-1952977854.

**Comment by yarikoptic** [comment]
is there a way to control how frequently it writes DB in persistent mode?  if we make it once a minute instead of once a second -- might be a better solution may be?

**Comment by candleindark** [comment]
> is there a way to control how frequently it writes DB in persistent mode? if we make it once a minute instead of once a second -- might be a better solution may be?

@yarikoptic You are right. In fact, I used the environment variable to control how frequently the writes to the DB happens in our Docker compose file. I totally forgot about it even though I made a mental note that the setting of the environment variable should be adjusted depending on performance.

I initially set the interval to be just 1 second because I was greedy and assumed that the writing would only include the difference in the state of Flower.

PR #307 increases the interval to 15 minutes. Let's see if that increase is sufficient after a few days when the DB is bigger.
